tka/sig.go

```diff
@@ -97,18 +133,56 @@ func (s *NodeKeySignature) Unserialize(data []byte) error {
 	return dec.Unmarshal(data, s)
 }
 
-// verifySignature checks that the NodeKeySignature is authentic and certified
-// by the given verificationKey.
-func (s *NodeKeySignature) verifySignature(verificationKey Key) error {
+// verifySignature checks that the NodeKeySignature is authentic, certified
+// by the given verificationKey, and authorizes the given nodeKey.
+func (s *NodeKeySignature) verifySignature(nodeKey key.NodePublic, verificationKey Key) error {
+	nodeBytes, err := nodeKey.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("marshalling pubkey: %v", err)
+	}
+	if !bytes.Equal(nodeBytes, s.Pubkey) {
+		return errors.New("signature does not authorize nodeKey")
+	}
+
 	sigHash := s.SigHash()
-	switch verificationKey.Kind {
-	case Key25519:
-		if ed25519consensus.Verify(ed25519.PublicKey(verificationKey.Public), sigHash[:], s.Signature) {
-			return nil
+	switch s.SigKind {
+	case SigRotation:
+		if s.Nested == nil {
+			return errors.New("nested signatures must nest a signature")
+		}
+
+		// Verify the signature using the nested rotation key.
+		verifyPub, ok := s.Nested.rotationPublic()
+		if !ok {
+			return errors.New("missing rotation key")
+		}
+		if !ed25519.Verify(ed25519.PublicKey(verifyPub[:]), sigHash[:], s.Signature) {
+			return errors.New("invalid signature")
+		}
+
+		// Recurse to verify the signature on the nested structure.
+		var nestedPub key.NodePublic
+		if err := nestedPub.UnmarshalBinary(s.Nested.Pubkey); err != nil {
+			return fmt.Errorf("nested pubkey: %v", err)
+		}
+		if err := s.Nested.verifySignature(nestedPub, verificationKey); err != nil {
```

Comment:
How well does Go handle deep recursive function calls? For tailnets that require daily key rotation, I could see this easily being hundreds or even thousands of levels deep.

Reply:
I think it has no issue with tens of thousands but it does cap it eventually. In our case, we'll hit our CBOR nesting limit well before that.